### ISSUE TITLE
FormValidateInput type doesn't include `undefined` on callback for `number` input

### ISSUE THREAD

**Issue by jamietre** [mantinedev/mantine#2973]
### What package has an issue

@mantine/form

### Describe the bug

When creating validation for a form, fields of type `number` are consumed by the validator using the same type (number), but the value `undefined` can be passed to the validator if the user deletes the default value.

This can result in bugs because of incorrect assumptions about the type of data that can be passed to the validator. The form validation example at [https://mantine.dev/form/validation/](https://mantine.dev/form/validation/) actually exhibits this problem - if you delete "0" for age, the form will validate successfully, since `undefined < [any number]` always evaluates to false.

Trivial example:
```
  type InputForm = {
    age: number;
  };

  const inputValidation: FormValidateInput<InputForm> = {
    age: (value) => {
      // "value" is typed as `number`, but can be `undefined`
      console.log(value);
     // this test (also used in the Matine web site example) will fail if `value === undefined`
      return value < 18 ? 'You must be at least 18 to register' : null;
    },
  };

  const form = useForm<InputForm>({
    initialValues: { age: 0 },
    validate: inputValidation,
  });
```

### What version of @mantine/hooks page do you have in package.json?

5.7.2

### If possible, please include a link to a codesandbox with the reproduced problem

https://codesandbox.io/s/keen-faraday-e3oynu?file=/package.json

### Do you know how to fix the issue

Yes

### Are you willing to participate in fixing this issue and create a pull request with the fix

No

### Possible fix

Possible solution is to make a special case for number, e.g. altering `FormRule` to map number to `Value | undefined`. However, I have fairly limited maintine experience, and I'm not sure if there are other possible situations where the type from a form could be widened in the validation.

I'd normally volunteer to make a PR but I have very limited experience with Mantine and don't know if this issue affects other types, or if this solution could have other consequences I'm not considering.

```
type PossiblyOptionalValue<T> = T extends number ? T | undefined : T;
type FormRule<Value, Values> = Value extends Array<infer ListValue>
    ?
          | Partial<
                {
                    [Key in keyof ListValue]: ListValue[Key] extends Array<
                        infer NestedListItem
                    >
                        ?
                              | FormRulesRecord<NestedListItem>
                              | Rule<ListValue[Key], Values>
                        :
                              | FormRulesRecord<ListValue[Key]>
                              | Rule<ListValue[Key], Values>;
                }
            >
          | Rule<Value, Values>
    : Value extends Record<string, unknown>
    ? FormRulesRecord<Value> | Rule<Value, Values>
    : Rule<PossiblyOptionalValue<Value>, Values>;
```

**Comment by rtivital** [comment]
You should define correct type on your side, use `number | undefined` instead – https://codesandbox.io/s/youthful-shannon-ficxvk
See docs – https://mantine.dev/form/values/#set-values-type

**Comment by jamietre** [comment]
Why wouldn't we want Mantine to provide the correct types automatically if it can? 

In any case, if this isn't considered a bug, then probably at least fix the example to account for age possibly being undefined ;) but this seems like a gotcha to me that could be improved with better types.

**Comment by rtivital** [comment]
Because `@mantine/form` package is meant to be used with any kind of inputs – even though you use it with `NumberInput` component which value is `number | undefined`, you can use it with component that accepts only `number` value – then the type will be incorrect for this component without any option of fixing it.

**Comment by rtivital** [comment]
Yes, it is how typescript infers values sometimes, no idea how to fix that, you can defined type on your side – https://codesandbox.io/s/youthful-shannon-ficxvk?file=/src/App.tsx

**Comment by jamietre** [comment]
> even though you use it with NumberInput component which value is number | undefined, you can use it with component that accepts only number value

I'd argue that if you have to choose between the type being too wide or too narrow, too wide is a much better option, since the former can result in bugs really easily. That's how I ended up here and found the same bug on the example I mentioned from the mantine docs!

**Comment by jamietre** [comment]
Possible solution to the specific problem of mapping types that include `null | undefined` is to add `NonNullable` to the `extends Array` test. (I'm not sure either why TS doesn't just fail that `extends` test instead of  mapping the whole thing to "any", but this ought to achieve the goal). 

```
declare type FormRule<Value, Values> = NonNullable<Value> extends Array<infer ListValue>
    ?
          | Partial<
                {
                    [Key in keyof ListValue]: NonNullable<ListValue[Key]> extends Array<
                        infer NestedListItem
                    >
                        ?
                              | FormRulesRecord<NestedListItem>
                              | Rule<ListValue[Key], Values>
                        :
                              | FormRulesRecord<ListValue[Key]>
                              | Rule<ListValue[Key], Values>;
                }
            >
          | Rule<Value, Values>
    : Value extends Record<string, unknown>
    ? FormRulesRecord<Value> | Rule<Value, Values>
    : Rule<Value, Values>;

**Comment by rtivital** [comment]
I'll try to implement that, thanks for investigation!

**Comment by rtivital** [comment]
Fixed in 5.9.1
